### PR TITLE
[MON-1871] Replace "app" selector with "app.kubernetes.io/name" for Prometheus Operator

### PR DESCRIPTION
**Pull request by raptorsun**
Prometheus-operator has[ deprecated the "app" selector in favor of "app.kubernetes.io/name"](https://github.com/prometheus-operator/prometheus-operator/blob/master/CHANGELOG.md#0480--2021-05-19). This pull request replaces the "app" selector with "app.kubernetes.io/name" when applicable.

[Link to JIRA issue](https://issues.redhat.com/browse/MON-1871)

We remove "app" selector label from upstream kube-prometheus in[ this PR](https://github.com/prometheus-operator/kube-prometheus/pull/1420).

* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: assets/alertmanager/service.yaml [hunk 1]
```diff
@@ -21,7 +21,6 @@ spec:
     targetPort: tenancy
   selector:
     alertmanager: main
-    app: alertmanager
```
Comment (fpetkovski) [inline assets/alertmanager/service.yaml:24]: Should we remove the label from the Alertmanager instance as well?

File: assets/prometheus-k8s/service.yaml [hunk 2]
```diff
@@ -20,7 +20,6 @@ spec:
     port: 9092
     targetPort: tenancy
   selector:
-    app: prometheus
```
Comment (fpetkovski) [inline assets/prometheus-k8s/service.yaml:23]: Shuold we change or remove this label in the prometheus object?

File: assets/prometheus-user-workload/service.yaml [hunk 3]
```diff
@@ -20,7 +20,6 @@ spec:
     port: 10902
     targetPort: thanos-proxy
   selector:
-    app: prometheus
```
Comment (fpetkovski) [inline assets/prometheus-user-workload/service.yaml:23]: Same as above

File: jsonnet/components/alertmanager.libsonnet [hunk 4]
```diff
@@ -88,6 +88,9 @@ function(params)
           },
         ],
         type: 'ClusterIP',
+        selector+: {
+          app:: null,
```
Comment (simonpasquier) [inline jsonnet/components/alertmanager.libsonnet:92]: shouldn't this be done in kube-prometheus directly?

File: assets/alertmanager/service.yaml (hunk 1 above)
Comment (raptorsun) [reply assets/alertmanager/service.yaml:24]: yes, this need be removed, too

File: jsonnet/components/alertmanager.libsonnet (hunk 4 above)
Comment (raptorsun) [reply jsonnet/components/alertmanager.libsonnet:92]: agree, we had better change it in kube-prometheus.
i'm going to make a pull request to kube-prometheus on this modification. meanwhile, we keep it here until upstream changes.

File: assets/alertmanager/service.yaml (hunk 1 above)
Comment (simonpasquier) [reply assets/alertmanager/service.yaml:24]: Not sure about this comment. Which label are we talking about?

Comment (fpetkovski) [reply assets/alertmanager/service.yaml:24]: I assume there is an `app` label on the alertmanager pods as well. If we are removing it from the selector, should we also remove it from the pods?

Comment (fpetkovski) [reply assets/alertmanager/service.yaml:24]: Understood, in that case this should fix itself in the future.

File: assets/cluster-monitoring-operator/service.yaml [hunk 5]
```diff
@@ -14,4 +14,4 @@ spec:
     port: 8443
     targetPort: https
   selector:
-    app: cluster-monitoring-operator
+    app.kubernetes.io/name: cluster-monitoring-operator
```
[the remark this answers is not in the document]
Comment (raptorsun) [reply assets/cluster-monitoring-operator/service.yaml:17]: the label in CMO deployment has been changed now

[the remark this answers is not in the document]
Comment (raptorsun) [reply assets/cluster-monitoring-operator/service.yaml:17]: changed now :)

**Comment by raptorsun** [comment]
Side note: this snippets is used to find out "app" labels to remove.

``` bash
find assets/ -type f -name "*.yaml" | while read -r f; do
    selector=$(tmp/bin/gojsontoyaml -yamltojson <"$f" | jq '.spec.selector')
    if [ "$selector" = "null" ]; then
        # echo $f has no .spec.selector
        continue
    fi

    app=$(echo $selector | jq -r '.app')

    if [ "$app" = "null" ]; then
        matchLabels=$(echo $selector | jq '.matchLabels')

        if [ "$matchLabels" = "null" ]; then
            # echo $f: has no .spec.selector.matchLabels
            continue
        else
            # echo $f: .spec.selector.matchLabels
            matchLabelsApp=$(echo $matchLabels | jq -r '.app')
            if [ "$matchLabelsApp" != "null" ]; then
                echo $f: .spec.selector.matchLabels.app=$matchLabelsApp
            fi
        fi

    else

        echo $f: .spec.selector.app=$app
    fi

done
```

File: manifests/0000_50_cluster-monitoring-operator_05-deployment.yaml [hunk 6]
```diff
@@ -5,6 +5,7 @@ metadata:
   namespace: openshift-monitoring
   labels:
     app: cluster-monitoring-operator
```
Comment (simonpasquier) [inline manifests/0000_50_cluster-monitoring-operator_05-deployment.yaml:7]: it would be good to remove this label while we're at it

Comment (raptorsun) [reply manifests/0000_50_cluster-monitoring-operator_05-deployment.yaml:7]: The selector at line 13-15 is immutable during upgrade, matching label `app: cluster-monitoring-operator`.
If we remove this label at line 7 and line 19, the CMO deploy may not find any pods to manage. So I leave both app labels where they are 😅
Test with removing them has been done with [this commit](https://github.com/openshift/cluster-monitoring-operator/pull/1388/files/40f7f8803dc24c752619c4abd65aa7df8577eea7#diff-984d250e7b24f523400c14314267a47c0c8fa79c62f07bef798af2c462f91777). The upgrade test did not pass.

**Comment by juzhao** [comment]
@raptorsun 
```
1. deploy cluster-monitoring-operator, metadata.labels has "app: cluster-monitoring-operator", same for spec.template.metadata.labels
2. sts alertmanager-main, spec.selector.matchLabels has "app: alertmanager" and "app.kubernetes.io/name: alertmanager" labels,  which "app: alertmanager" should be removed
3. sts prometheus-k8s, spec.selector.matchLabels has "app: prometheus" and "app.kubernetes.io/name: prometheus" labels, which "app: prometheus" should be removed
4. UWM sts prometheus-user-workload, spec.selector.matchLabels  has "app: prometheus" and "app.kubernetes.io/name: prometheus" labels, which "app: prometheus"  should be removed
5. UWM sts thanos-ruler-user-workload, spec.selector.matchLabels  has "app: thanos-ruler" and "app.kubernetes.io/name: thanos-ruler" labels, which "app: thanos-ruler"  should be removed
```
see http://pastebin.test.redhat.com/1002649

**Comment by juzhao** [comment]
based on the previous comment, set /lgtm

**Pull request merged**